File: android/sdks/zephr_agent_runtime/src/main/java/xyz/zephr/sdks/agent/litertlm/0_LiteRtLmAPI.kt
package xyz.zephr.sdks.agent.litertlm

import xyz.zephr.sdks.agent.ZephrAgentRuntime

/**
 * LiteRT-LM shaped public facade over the Zephr Agent runtime.
 *
 * This file intentionally exposes top-level API names so callers can use:
 *
 * import xyz.zephr.sdks.agent.litertlm.*
 */

public typealias Engine = ZephrAgentRuntime.Conversation.Engine

public typealias Conversation = ZephrAgentRuntime.Conversation.Conversation

public typealias ConversationConfig = ZephrAgentRuntime.Conversation.Config

public typealias Contents = ZephrAgentRuntime.Conversation.Contents

public typealias Message = ZephrAgentRuntime.Conversation.Message

public typealias Strategy = ZephrAgentRuntime.Conversation.Strategy

public typealias Tool = ZephrAgentRuntime.Conversation.Tool

public typealias ToolSet = ZephrAgentRuntime.Conversation.ToolSet

public typealias OpenApiTool = ZephrAgentRuntime.Conversation.OpenApiTool

public typealias ToolProvider = ZephrAgentRuntime.Conversation.ToolProvider

public typealias ModelChannel = ZephrAgentRuntime.Lifecycle.ModelChannel

public data class EngineConfig(
    val modelChannel: ModelChannel = ModelChannel.PUBLIC,
    val textBackend: Backend = Backend.GPU,
    val embeddingBackend: Backend = Backend.OFF,
    val vlmBackend: Backend = Backend.OFF,
    val litertCompilationCacheEnabled: Boolean = false,
    val litertRuntimeLibraryDir: String? = null,
    val gemma4Runtime: ZephrAgentRuntime.Lifecycle.Gemma4Options = ZephrAgentRuntime.Lifecycle.Gemma4Options(),
    val diagnosticGemma4: ZephrAgentRuntime.Diagnostics.Gemma4Options = ZephrAgentRuntime.Diagnostics.Gemma4Options(),
    val numThreads: Int = 0,
    val llmExecutionChoiceId: String = "",
    val ragEmbeddingExecutionChoiceId: String = "",
    val vlmExecutionChoiceId: String = "",
) {
    internal fun toLifecycleConfiguration(): ZephrAgentRuntime.Lifecycle.Configuration =
        ZephrAgentRuntime.Lifecycle.Configuration(
            modelChannel = modelChannel,
            llmExecutionChoiceId = llmExecutionChoiceId.ifBlank {
                textChoiceId(textBackend)
            },
            ragEmbeddingExecutionChoiceId = ragEmbeddingExecutionChoiceId.ifBlank {
                embeddingChoiceId(embeddingBackend)
            },
            vlmExecutionChoiceId = vlmExecutionChoiceId.ifBlank {
                visionChoiceId(vlmBackend)
            },
            litertCompilationCacheEnabled = litertCompilationCacheEnabled,
            litertRuntimeLibraryDir = litertRuntimeLibraryDir,
            gemma4Runtime = gemma4Runtime,
            diagnosticGemma4 = diagnosticGemma4,
            numThreads = numThreads,
        )
}

public enum class Backend {
    CPU,
    GPU,
    NPU,
    OFF;
}

public suspend fun Engine.initialize(configuration: EngineConfig) {
    initialize(configuration.toLifecycleConfiguration())
}

public suspend fun Engine.prepare(configuration: EngineConfig) {
    prepare(configuration.toLifecycleConfiguration())
}

public fun tool(toolSet: ToolSet): ToolProvider =
    ZephrAgentRuntime.Conversation.tool(toolSet)

public fun tool(openApiTool: OpenApiTool): ToolProvider =
    ZephrAgentRuntime.Conversation.tool(openApiTool)

private fun textChoiceId(backend: Backend): String =
    when (backend) {
        Backend.CPU -> "gemma4.cpu"
        Backend.GPU -> "gemma4.gpu"
        Backend.NPU -> "gemma4.npu"
        Backend.OFF -> "off"
    }

private fun embeddingChoiceId(backend: Backend): String =
    when (backend) {
        Backend.CPU -> "embedding.gemma3.cpu"
        Backend.GPU -> "embedding.gemma3.gpu"
        Backend.NPU -> "embedding.gemma3.npu"
        Backend.OFF -> "off"
    }

private fun visionChoiceId(backend: Backend): String =
    when (backend) {
        Backend.CPU -> "gemma4_vision.cpu"
        Backend.GPU -> "gemma4_vision.gpu"
        Backend.NPU -> "gemma4_vision.npu"
        Backend.OFF -> "off"
    }

File: apple/sdks/zephr_agent_runtime/Sources/0_LiteRTLMAPI.swift
import Foundation

extension ZephrAgentRuntime {
    public enum LiteRTLM {}
}

extension ZephrAgentRuntime.LiteRTLM {

public typealias Engine = ZephrAgentRuntime.Conversation.Engine
public typealias Conversation = ZephrAgentRuntime.Conversation.Conversation
public typealias ConversationConfig = ZephrAgentRuntime.Conversation.Config
public typealias Contents = ZephrAgentRuntime.Conversation.Contents
public typealias Message = ZephrAgentRuntime.Conversation.Message
public typealias Tool = ZephrAgentRuntime.Conversation.Tool
public typealias ToolParam = ZephrAgentRuntime.Conversation.ToolParam
public typealias ToolParameterValue = ZephrAgentRuntime.Conversation.ToolParameterValue
public typealias ToolProvider = ZephrAgentRuntime.Conversation.ToolProvider
public typealias OpenApiTool = ZephrAgentRuntime.Conversation.OpenApiTool
public typealias ModelChannel = ZephrAgentRuntime.Lifecycle.ModelChannel

public enum Backend: Sendable {
    case cpu
    case gpu
    case npu
    case off
}

public struct EngineConfig: Sendable {
    public var modelChannel: ModelChannel
    public var textBackend: Backend
    public var embeddingBackend: Backend
    public var vlmBackend: Backend
    public var litertCompilationCacheEnabled: Bool
    public var gemma4Runtime: ZephrAgentRuntime.Lifecycle.Gemma4Options
    public var diagnosticGemma4: ZephrAgentRuntime.Diagnostics.Gemma4Options
    public var numThreads: Int
    public var llmExecutionChoiceID: String
    public var ragEmbeddingExecutionChoiceID: String
    public var vlmExecutionChoiceID: String
    public var modelStorage: ZephrAgentRuntime.Lifecycle.ModelStorage?

    public init(
        modelChannel: ModelChannel = .public,
        textBackend: Backend = .gpu,
        embeddingBackend: Backend = .off,
        vlmBackend: Backend = .off,
        litertCompilationCacheEnabled: Bool = false,
        gemma4Runtime: ZephrAgentRuntime.Lifecycle.Gemma4Options = .automatic,
        diagnosticGemma4: ZephrAgentRuntime.Diagnostics.Gemma4Options = .disabled,
        numThreads: Int = 0,
        llmExecutionChoiceID: String = "",
        ragEmbeddingExecutionChoiceID: String = "",
        vlmExecutionChoiceID: String = "",
        modelStorage: ZephrAgentRuntime.Lifecycle.ModelStorage? = nil
    ) {
        self.modelChannel = modelChannel
        self.textBackend = textBackend
        self.embeddingBackend = embeddingBackend
        self.vlmBackend = vlmBackend
        self.litertCompilationCacheEnabled = litertCompilationCacheEnabled
        self.gemma4Runtime = gemma4Runtime
        self.diagnosticGemma4 = diagnosticGemma4
        self.numThreads = numThreads
        self.llmExecutionChoiceID = llmExecutionChoiceID
        self.ragEmbeddingExecutionChoiceID = ragEmbeddingExecutionChoiceID
        self.vlmExecutionChoiceID = vlmExecutionChoiceID
        self.modelStorage = modelStorage
    }

    public var lifecycleConfiguration: ZephrAgentRuntime.Lifecycle.Configuration {
        ZephrAgentRuntime.Lifecycle.Configuration(
            modelChannel: modelChannel,
            llmExecutionChoiceID: llmExecutionChoiceID.isEmpty ? Self.textChoiceID(textBackend) : llmExecutionChoiceID,
            ragEmbeddingExecutionChoiceID: ragEmbeddingExecutionChoiceID.isEmpty
                ? Self.embeddingChoiceID(embeddingBackend)
                : ragEmbeddingExecutionChoiceID,
            vlmExecutionChoiceID: vlmExecutionChoiceID.isEmpty ? Self.visionChoiceID(vlmBackend) : vlmExecutionChoiceID,
            litertCompilationCacheEnabled: litertCompilationCacheEnabled,
            gemma4Runtime: gemma4Runtime,
            diagnosticGemma4: diagnosticGemma4,
            numThreads: numThreads,
            modelStorage: modelStorage
        )
    }

    private static func textChoiceID(_ backend: Backend) -> String {
        switch backend {
        case .cpu:
            "gemma4.cpu"
        case .gpu:
            "gemma4.gpu"
        case .npu:
            "gemma4.npu"
        case .off:
            "off"
        }
    }

    private static func embeddingChoiceID(_ backend: Backend) -> String {
        switch backend {
        case .cpu:
            "embedding.gemma3.cpu"
        case .gpu:
            "embedding.gemma3.gpu"
        case .npu:
            "embedding.gemma3.npu"
        case .off:
            "off"
        }
    }

    private static func visionChoiceID(_ backend: Backend) -> String {
        switch backend {
        case .cpu:
            "gemma4_vision.cpu"
        case .gpu:
            "gemma4_vision.gpu"
        case .npu:
            "gemma4_vision.npu"
        case .off:
            "off"
        }
    }
}

public static func tool<T: Tool>(_ tool: T) -> ToolProvider {
    ZephrAgentRuntime.Conversation.tool(tool)
}

public static func tool(_ openApiTool: any OpenApiTool) -> ToolProvider {
    ZephrAgentRuntime.Conversation.tool(openApiTool)
}

}

public extension ZephrAgentRuntime.LiteRTLM.Engine {
    func initialize(
        configuration: ZephrAgentRuntime.LiteRTLM.EngineConfig,
        progress: ZephrAgentRuntime.Lifecycle.ModelProgressHandler? = nil
    ) async throws {
        try await initialize(configuration: configuration.lifecycleConfiguration, progress: progress)
    }

    func prepare(
        configuration: ZephrAgentRuntime.LiteRTLM.EngineConfig,
        progress: ZephrAgentRuntime.Lifecycle.ModelProgressHandler? = nil
    ) async throws {
        try await prepare(configuration: configuration.lifecycleConfiguration, progress: progress)
    }
}
